### PR TITLE
Adds new schedule- and pointer-aware local storage transformation

### PR DESCRIPTION
**Pull request by lukastruemper**
## Description
Creates a new LocalStorage transformation (existing ones to be deprecated later), which must be applied after the loop nest has been scheduled. The existing LocalStorage transformations are correct on a dataflow level, but need to guess the later schedule which is almost always incorrect (cooperative vs. thread-local). The PR adds comprehensive unit tests (~3000 LOC).

## Related Issues

- Closes #

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactor / internal change (no functional change)
- [ ] Documentation update

## Quality Checklist
*Please ensure the following are completed before requesting review:*

* [ ] Runtime-compatible (no externally visible existing method has its parameters changed, was deleted or has its exception behavior changed)
* [ ] Compile-time compatible (new arguments & properties added have default values, but users need to recompile against the new version to work)

IR:
* [ ] Introduces new types into the SDFG IR
* [ ]  Changed / added properties on SDFG IR Elements (LibraryNodes, ControlFlowNodes DataFlowNodes etc.)
  *  [ ] New / modified types have serialization / deserialization support
  *  [ ] Deserialization is backward compatible (new serialization can deserialize old format and represent it in the new format if applicable)
  *  [ ] Serialized format is backward compatible (previous deserializer can read new serialized format without incorrectness (just lacking additional guarantees that are optional or did not exist before)
  *  [ ] Serializer has an option to emit the previous format

### Unit Tests
- [ ] New unit tests have been added for the changes.
- [ ] Existing unit tests pass locally.
- [ ] Edge cases and error paths are covered.

### Integration Tests
- [ ] New integration tests have been added (or existing ones updated).
- [ ] Integration tests pass locally.
- [ ] Interactions with other components/services are verified.

### Documentation
- [ ] Public APIs, options, and behavior changes are documented.
- [ ] README / docs site / inline comments updated where applicable.
- [ ] Changelog / release notes updated (if applicable).

### Test Coverage
- [ ] Test coverage has not decreased as a result of this change.
- [ ] Coverage report reviewed and acceptable.
- [ ] Critical / new code paths are covered.

## How Has This Been Tested?
Describe the tests you ran to verify your changes and provide instructions to reproduce.
Include relevant details for your test configuration (OS, Python/LLVM/Docc version, hardware/target).

## Additional Context
Add any other context about the PR here (design decisions, trade-offs, follow-ups).

### CODE REVIEW

File: opt/src/transformations/local_storage.cpp [hunk 1]
```diff
@@ -0,0 +1,835 @@
...
+#include "sdfg/types/type.h"
+
+namespace sdfg {
+namespace transformations {
+
+namespace {
+
+/// Visit every Block reachable under @p node (recursing through sequences,
+/// loops, and if-else branches).
+void for_each_block(
```
Comment (Atrisan) [inline opt/src/transformations/local_storage.cpp:36]: Nitpick: Not sure if I am a fan of recursive higher-order function instead of a proper visitor, like the one defined ~50 lines later.

File: opt/src/transformations/local_storage.cpp [hunk 2]
```diff
@@ -0,0 +1,835 @@
...
+#include "sdfg/targets/gpu/gpu_schedule_type.h"
+#include "sdfg/types/array.h"
+#include "sdfg/types/pointer.h"
+#include "sdfg/types/scalar.h"
+#include "sdfg/types/type.h"
+
+namespace sdfg {
+namespace transformations {
+
+namespace {
```
Comment (Atrisan) [inline opt/src/transformations/local_storage.cpp:32]: Nitpick: Why the sub-namespace?

File: opt/src/transformations/local_storage.cpp [hunk 3]
```diff
@@ -0,0 +1,835 @@
...
+    void on_write_via(const std::string& c, const structured_control_flow::ControlFlowNode*, const data_flow::Memlet*) {
+        if (c == container) writes = true;
+    }
+};
+
+/// Composes the shared PointerEscape/Overwrite/Used analyzers over a subtree,
+/// mirroring MemoryOwnershipAnalysis. Adds one refinement DataDependencyAnalysis
+/// carries but the analyzers do not: a library node consuming the pointer with a
+/// missing or non-`no_capture` `pointer_access_type` is treated as aliasing.
+class ContainerAccessVisitor : public analysis::BaseUserVisitor,
```
Comment (Atrisan) [inline opt/src/transformations/local_storage.cpp:80]: Nitpick: Shouldn't this definition be moved into the header?

File: opt/src/transformations/local_storage.cpp [hunk 4]
```diff
@@ -0,0 +1,835 @@
...
+        if (!is_gpu && sched.category() == structured_control_flow::ScheduleTypeCategory::None) {
+            continue;
+        }
+        LocalityPlan::Dim d;
+        d.indvar = sloop->indvar();
+        d.is_gpu = is_gpu;
+        d.cooperative = is_cooperative(d.indvar);
+        if (is_gpu) {
+            const std::string& value = sched.value();
+            if (value == "CUDA_Offload" || value == "ROCM_Offload") {
```
Comment (Atrisan) [inline opt/src/transformations/local_storage.cpp:391]: Nitpick: Why not use is_gpu_schedule(const structured_control_flow::ScheduleType& schedule) from gpu_map_utils instead?
Or do you expect issues with the old schedule types?

File: opt/src/transformations/local_storage.cpp [hunk 5]
```diff
@@ -0,0 +1,835 @@
...
+                    break;
+                }
+            }
+            if (!coop_map || !symbolic::eq(coop_map->indvar(), coop_dims.front().indvar)) {
+                return false;
+            }
+            // The cooperative copy is lowered by the new offload dispatcher, so
+            // only the *_Offload schedules are supported (not the legacy ones).
+            const std::string& sched_value = coop_map->schedule_type().value();
+            if (sched_value != "CUDA_Offload" && sched_value != "ROCM_Offload") {
```
Comment (Atrisan) [inline opt/src/transformations/local_storage.cpp:590]: Nitpick: S.o. !is_gpu_schedule()?

**Review by Atrisan — APPROVED** [review]
Besides a few nitpicks, LGTM.

File: opt/src/transformations/local_storage.cpp (hunk 1 above)
Comment (lukastruemper) [reply opt/src/transformations/local_storage.cpp:36]: I copied this from another pass, we can probably unify this later

File: opt/src/transformations/local_storage.cpp (hunk 3 above)
Comment (lukastruemper) [reply opt/src/transformations/local_storage.cpp:80]: It's currently not ideal. Ramon added the Pointeranalyzers but we haven't really factored them into easy to use analysis, so I keep them private/internal utils

File: opt/src/transformations/local_storage.cpp (hunk 2 above)
Comment (lukastruemper) [reply opt/src/transformations/local_storage.cpp:32]: Readability / collapsing

File: opt/src/transformations/local_storage.cpp (hunk 4 above)
Comment (lukastruemper) [reply opt/src/transformations/local_storage.cpp:391]: Yes, the branch for the old one is below

File: opt/src/transformations/local_storage.cpp (hunk 5 above)
Comment (lukastruemper) [reply opt/src/transformations/local_storage.cpp:590]: Yes, the old is treated separately
